src/charm.py

```diff
@@ -0,0 +1,301 @@
+#!/usr/bin/env python3
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import glob
+import time
+from typing import Optional
+
+import logging
+from pathlib import Path
+
+from ops.charm import CharmBase
+from ops.main import main
+from ops.model import ActiveStatus, WaitingStatus, MaintenanceStatus, BlockedStatus
+
+import lightkube
+from lightkube import codecs
+from lightkube.resources.apps_v1 import StatefulSet
+from lightkube.core.exceptions import ApiError
+
+
+METACONTROLLER_IMAGE = "metacontroller/metacontroller:v0.3.0"
+
+
+class MetacontrollerOperatorCharm(CharmBase):
+    """Charm the Metacontroller"""
+
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        if not self.unit.is_leader():
+            self.model.unit.status = WaitingStatus("Waiting for leadership")
+            return
+
+        self.framework.observe(self.on.install, self._install)
+        self.framework.observe(self.on.remove, self._remove)
+        self.framework.observe(self.on.update_status, self._update_status)
+        # self.framework.observe(self.on.config_changed, self._reconcile)
+
+        self.logger = logging.getLogger(__name__)
+
+        # TODO: Fix file imports and move ./src/files back to ./files
+        self._manifests_file_root = None
+        self.manifest_file_root = "./src/files/manifests/"
```

Comment:
This is very confusing, they are named mostly the same, but they point to different paths and one is defined somewhere down below. Not sure if we really need them, or if we can just use pathlib to define these.

Reply:
I think the confusion here is that `self.manifest_file_root` is a property (see [here](https://github.com/canonical/metacontroller-operator/blob/85c3af536c8afbf2ac292e922fbfb7591930c950/src/charm.py#L192)).  It is done this way so that it forces everything into a `pathlib.Path` as I agree with you they're nicer to use :)

Reply:
I agree with @DnPlas this is overly complicated. This code is controlled by the charm, you don't have to set so many guardrails to ensure things are set correctly. I would set this path with pathlib once as an attribute and move on with it.

Reply:
After talking with @DomFleischmann I have relented and removed the properties :)  My vote is still that the complexity is minor and the robustness is nice, but I wont die on this one

To make my conscious feel better, I added type hints to all the attributes, that way we nudge people toward a `Path` instead of a `str`, etc.